bitbake-setup: sub operations for `settings`

js731ca opened the pull request:
Previously the sub-command 'settings' would take any number of
arguments and then silently do nothing if the number wasn't three.

The help text was also not clear about this, marking the positionals
separately as optional:

`usage: bitbake-setup settings [-h] [--global] [--unset UNSET UNSET] [-l] [section] [key] [value]`

The '--unset SECTION SETTING' also did not  integrate too well, as it had its own positional arguments for section+setting.

For a bit more consistency and a explorable help, a sub-subparser is added, that provides the commands:
```
  bitbake-setup settings list
  bitbake-setup settings set foo bar baz
  bitbake-setup settings unset foo bar
```
with a '--global' that is added from a stand-alone parent parser, so
that it shows up in all sub-command help texts.

The new help text now reads:
`usage: bitbake-setup settings [-h] [--global] {list,set,unset} ...`

and the respective sub commands:
```
usage: bitbake-setup settings list [-h] [--global]
usage: bitbake-setup settings set [-h] [--global] <section> <setting> <value>
usage: bitbake-setup settings unset [-h] [--global] <section> <setting>
```

js731ca committed:
bitbake-setup: add 'metavar' for self-descriptive parameters

Add a metavar to the argparse options to have a self-descriptive help text.
Otherwise argpase defaults to use the argument name in all-uppercase.

Before:

usage: bitbake-setup [-h] [-d] [-q] [--color COLOR] [--no-network] [--global-settings GLOBAL_SETTINGS] [--setting SETTING SETTING SETTING]
                     {list,init,status,update,install-buildtools,settings} ...

After:

usage: bitbake-setup [-h] [-d] [-q] [--color COLOR] [--no-network] [--global-settings PATH] [--setting SECTION SETTING VALUE]
                     {list,init,status,update,install-buildtools,settings} ...

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca committed:
bitbake-setup: use args.cmdline_settings for --settings

To align the storage destination with the internally used variable
name. This makes room for having another option use 'args.setting'

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca committed:
bitbake-setup: commandline: use subsubparser for settings {list,set,unset}

Previously the sub-command 'settings' would take any number of
arguments and then silently do nothing if the number wasn't three.

The help text was also not clear about this, marking the positionals
separately as optional:

usage: bitbake-setup settings [-h] [--global] [--unset UNSET UNSET] [-l] [section] [key] [value]

The '--unset SECTION SETTING' also did not  integrate too well, as it
had its own positional arguments for section+setting.

For a bit more consistency and a explorable help, a sub-subparser is
added, that provides the commands:
  bitbake-setup settings list
  bitbake-setup settings set foo bar baz
  bitbake-setup settings unset foo bar
with a '--global' that is added from a stand-alone parent parser, so
that it shows up in all sub-command help texts.

The new help text now reads:
usage: bitbake-setup settings [-h] [--global] {list,set,unset} ...

and the respective sub commands:
usage: bitbake-setup settings list [-h] [--global]
usage: bitbake-setup settings set [-h] [--global] <section> <setting> <value>
usage: bitbake-setup settings unset [-h] [--global] <section> <setting>

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

kanavin commented:
I like this idea, but

some people have expressed the wish for bitbake-setup to 'copy' 'git config' command line UI. 

In my opinion, the 'git config' UI is *awful*, but there's merit in reusing something many users already know. What you think?

kanavin commented:
> I like this idea, but
> 
> some people have expressed the wish for bitbake-setup to 'copy' 'git config' command line UI.
> 
> In my opinion, the 'git config' UI is _awful_, but there's merit in reusing something many users already know. What you think?

On the other hand, maybe we shouldn't go that far. Just having a 'stacked hierarchy' of settings, and a way to say --global when changing them is close enough to how git does it.

kanavin reviewed (changes requested):
Don't forget to fix the tests too :)

js731ca commented:
instead of a sub-subparser we could also make use of a mutually-exclusive group - which is more like git-config with all its '--[no-]option'(?)
But the usage would then look like this:
```
bitbake-setup settings --set foo bar baz
bitbake-setup settings --list
bitbake-setup settings --unset foo bar
```

Side question: should sth like this be supported?
```
bitbake-setup settings --set foo bar baz --set foo boo baa --unset foo bar
```

----
IMO the sub-subcommand with its explorable help is "nicer" :-)

kanavin commented:
> instead of a sub-subparser we could also make use of a mutually-exclusive group - which is more like git-config with all its '--[no-]option'(?) But the usage would then look like this:
> 
> ```
> bitbake-setup settings --set foo bar baz
> bitbake-setup settings --list
> bitbake-setup settings --unset foo bar
> ```
> 
> IMO the sub-subcommand with its explorable help is "nicer" :-)


Yes, let's stay with the sub-command rework.

kanavin commented:
We really do need to proceed with this first, as it breaks (again) the settings UI, and this really needs to be done ASAP, ideally before RP lands bitbake-setup in master.
So I'd suggest #14 is set aside for a moment and this is prioritized.

js731ca commented:
v2: adding tests,
picking up the metavar change from #15

kanavin commented on bin/bitbake-setup:
```diff
@@ -784,15 +782,29 @@ def main():
-    parser_settings.add_argument('--unset', nargs=2, help="Unset a setting, e.g. 'bitbake-setup settings --unset default registry' would revert to the registry setting in a global settings file")
-    parser_settings.add_argument('-l' ,'--list', action='store_true', help="List all settings with their values")
+    parser_settings_arg_global = argparse.ArgumentParser(add_help=False)
+    parser_settings_arg_global.add_argument('--global', action='store_true', help="Modify the setting in a global settings file, rather than one specific to a top directory")
```
--global should be added only to set and unset sub-commands, as it has no meaning in 'list'.

kanavin commented on lib/bb/tests/setup.py:
```diff
@@ -232,25 +232,25 @@ def test_setup(self):
+        out = self.runbbsetup("settings list")
         self.assertIn("default top-dir-prefix {}".format(self.tempdir), out[0])
         self.assertIn("default dl-dir {}".format(os.path.join(self.tempdir, 'downloads')), out[0])
         self.assertIn("default registry {}".format('git://{};protocol=file;branch=master;rev=master'.format(self.registrypath)), out[0])
```
Please squash this commit into the previous one, as having them separately breaks bisectability of bitbake-selftest.

js731ca commented:
v2.1: squashed tests into the commit that adds the subcommand
removed parents=parser_settings_arg_global  from the list command, to remove that option
